cirq-core/cirq/transformers/transformer_api.py

```diff
@@ -300,14 +264,45 @@ def transformer(cls_or_func: Any) -> Any:
     >>>        ...
 
     Args:
-        cls_or_func: The callable class or method to be decorated.
+        cls_or_func: The callable class or function to be decorated.
 
     Returns:
-        Decorated class / method which includes additional logging boilerplate. The decorated
-        callable always receives a copy of the input circuit so that the input is never mutated.
+        Decorated class / function which includes additional logging boilerplate.
     """
     if isinstance(cls_or_func, type):
-        return _transformer_class(cls_or_func)
+        cls = cls_or_func
+        method = cls.__call__
+
+        @functools.wraps(method)
+        def method_with_logging(self, circuit, context):
+            return _transform_and_log(
```

Comment:
`_transform_and_log` is now returning an `cirq.AbstractCircuit`. How is it getting converted to the possibly stricter return type of the input? 

Since we don't have any type annotations on this method, it doesn't check the return type of `_transform_and_log` and directly assumes that the untyped method preserve the input type?

Reply:
Added basic type annotations, and still checks, so mypy will give us a _bit_ more assurance that things are implemented correctly.